File: backend/src/routes/reports.js
import express from "express";
import { requireAuth } from "../middleware/auth.js";
import { ROLES } from "../constants/roles.js";
import { getAllUsers, getTeamMembers } from "../services/userService.js";
import { getMonthlyWfhUsage } from "../services/workLocationService.js";
import { getSettings } from "../services/settingsService.js";

const router = express.Router();

/**
 * GET /api/reports/wfh-overage?month=YYYY-MM
 * Returns only over-limit employees with rollup summary.
 * Employee → 403. TL → own team. Admin/Logistics → all.
 */
router.get("/wfh-overage", requireAuth, (req, res) => {
    const user = req.session.user;

    if (user.role === ROLES.EMPLOYEE) {
        return res.status(403).json({ error: "Access denied" });
    }

    const month = req.query.month || new Date().toISOString().slice(0, 7);
    const { monthlyWfhAllowance } = getSettings();

    let userList;
    if (user.role === ROLES.TEAM_LEAD) {
        userList = getTeamMembers(user.teamId);
    } else {
        userList = getAllUsers();
    }

    const allUsage = getMonthlyWfhUsage(userList, month, monthlyWfhAllowance);
    const overLimit = allUsage.filter((u) => u.overLimit);

    const summary = {
        overLimitCount: overLimit.length,
        totalExtraDays: overLimit.reduce((sum, u) => sum + (u.extraDays || 0), 0),
    };

    res.json({ month, allowance: monthlyWfhAllowance, summary, employees: overLimit });
});

export default router;

File: mhp-python/app/routes/reports.py
from fastapi import APIRouter, Depends, HTTPException
from datetime import date as DateType

from app.auth import get_current_user
from app.constants import Roles
from app.services.work_location_service import get_monthly_wfh_usage
from app.services.user_service import get_all_users, get_team_members
from app.services.settings_service import get_settings

router = APIRouter()


@router.get("/wfh-overage")
def wfh_overage(month: str, user: dict = Depends(get_current_user)):
    if user["role"] == Roles.EMPLOYEE:
        raise HTTPException(status_code=403, detail="Employees cannot view WFH overage reports")

    settings = get_settings()
    allowance = settings.get("monthlyWfhAllowance", 5)

    if user["role"] == Roles.TEAM_LEAD:
        user_list = get_team_members(user["teamId"])
    else:
        user_list = get_all_users()

    usage = get_monthly_wfh_usage(user_list, month, allowance)
    over = [u for u in usage if u["overLimit"]]

    return {
        "month": month,
        "allowance": allowance,
        "overageCount": len(over),
        "users": over,
    }
